Add GPIO interrupt support and embedded-hal-async trait implementation

MathisDeroo opened the pull request.

MathisDeroo committed:
Add GPIO interrupt support and embedded-hal-async trait implementation

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

felipebalbi commented on examples/src/bin/gpio_interrupt.rs:
```diff
@@ -0,0 +1,37 @@
+
+    unsafe {
+        cortex_m::interrupt::enable();
+    }
```
These are not supposed to be required. The GPIO driver should enable its own interrupts within the HAL. Look at [embassy-rp](https://github.com/embassy-rs/embassy/blob/main/embassy-rp/src/gpio.rs#L222-L233) for example

MathisDeroo replied:
Only hal::interrupt::GPIO1.enable(); should be handle by the GPIO HAL ?
cortex_m::interrupt::enable(); should be kept ?

felipebalbi replied:
no, no. Both should be handled internally :-)

felipebalbi commented on examples/src/bin/gpio_interrupt.rs:
```diff
@@ -0,0 +1,37 @@
+use hal::gpio::{DriveStrength, Input, Pull, SlewRate};
+use {defmt_rtt as _, embassy_mcxa as hal, panic_probe as _};
+
+use hal::interrupt::{InterruptExt};
```
Shouldn't be needed here

MathisDeroo replied:
I assume you are referring to use hal::interrupt::{InterruptExt}; only ?

felipebalbi replied:
Yes :-) The rest was added by github :-)

felipebalbi commented on examples/src/bin/button_async.rs:
```diff
@@ -0,0 +1,37 @@
+
+    let mut press_count = 0u32;
+
+    loop {
+        pin.wait_for_falling_edge().await;
+
+        press_count += 1;
+
+        defmt::info!("Button pressed! Count: {}", press_count);
+        Timer::after_millis(50).await;
+    }
```
Usually, if you look at other HALs, this example would be named `button_async.rs`. We may want to follow suit if only for consistency across HALs.

felipebalbi commented on src/gpio.rs:
```diff
@@ -3,13 +3,84 @@
+        }
+
+        INTERRUPT_DETECTED[port_index].store(true, Ordering::Relaxed);
+    }
```
For something like this, I very much like the `BitIter` pattern, which guarantees iteration only through set bits.

See more:

1. https://github.com/embassy-rs/embassy/blob/96a026c73bad2ebb8dfc78e88c9690611bf2cb97/embassy-imxrt/src/lib.rs#L160-L176
2. https://github.com/embassy-rs/embassy/blob/96a026c73bad2ebb8dfc78e88c9690611bf2cb97/embassy-stm32/src/exti.rs#L78-L92
3. https://github.com/embassy-rs/embassy/blob/96a026c73bad2ebb8dfc78e88c9690611bf2cb97/embassy-nrf/src/gpiote.rs#L253-L268
4. https://github.com/embassy-rs/embassy/blob/96a026c73bad2ebb8dfc78e88c9690611bf2cb97/embassy-nxp/src/lib.rs#L182-L196
5. https://github.com/embassy-rs/embassy/blob/96a026c73bad2ebb8dfc78e88c9690611bf2cb97/embassy-mspm0/src/lib.rs#L222-L236

felipebalbi commented on src/gpio.rs:
```diff
@@ -3,13 +3,84 @@
+#[interrupt]
+fn GPIO4() {
+    irq_handler(4, crate::pac::Gpio4::ptr());
+}
```
I think these need to be wrapped with `#[cfg(feature = "rt")]`.

felipebalbi commented on src/gpio.rs:
```diff
@@ -3,13 +3,84 @@
+    irq_handler(4, crate::pac::Gpio4::ptr());
+}
+
+
```
nit: one blank line is enough.

felipebalbi commented on src/gpio.rs:
```diff
@@ -3,13 +3,84 @@
+            }
+        }
+
+        INTERRUPT_DETECTED[port_index].store(true, Ordering::Relaxed);
```
Why do you need this? Isn't the status latched? It looks like you can clear the relevant ICR bit, but still read IFSR from the `poll` implementation. No?

felipebalbi commented on src/gpio.rs:
```diff
@@ -691,12 +792,97 @@ impl<'d> Input<'d> {
+                Poll::Ready(())
+            } else {
+                Poll::Pending
+            }
```
Here. If you read IFSR here, does it still contain the bit set? It seems like you could read that bit here, then clear it if it was set, then reenable ICR to allow any other interrupts to trigger.

MathisDeroo replied:
The issue I faced is that if I only rely only on the ICR[ISF] bit or IFSR register (ISF can be cleared with either register field per the documentation), then the bit is 0 until an interrupt is detected and set to 1 when detected.
If I Poll::Pending while the bit is 0 and Poll::Ready(()) when the bit is set to 1. When the IRQ handler will clear the bit ? Before of after the Poll:: change ? If it is done before, then in the poll() function the ICR[ISF] will never be 1

MathisDeroo replied:
After testing it seems that the IRQ is called before poll(), meaning if I clear the ICR[ISF] (or IFSR, as clearing one or the other clears both) in the IRQ then the bit in poll() will never be 1 and Poll:: remains Pending.
So I had to use the INTERRUPT_DETECTED mechanism to solve this limitation.
Do you have any other idea ?

felipebalbi replied:
Okay, it was a long shot. Thanks for testing. If there's no other way, let's keep you atomic u32 in place. Would be nice if we had separate per-pin status and mask bits for the interrupts

Commit:
Merge remote-tracking branch 'odp/main' into gpio_interrupt_support

Commit:
Run cargo fmt

MathisDeroo committed:
Improve GPIO driver interrupt mechanism and example

- GPIO interrupt managed internally at the HAL level,
- Renamed and cleaned gpio_interrupt example; now button_async.rs,
- Use BitIter instead of simple for loop in the irq handler,
- Fix comments and add "rt" wrappen to GPIO IRQ handler.

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
Merge branch 'gpio_interrupt_support' of https://github.com/MathisDeroo/embassy-mcxa into gpio_interrupt_support

felipebalbi commented on src/gpio.rs:
```diff
@@ -691,10 +821,92 @@ impl<'d> Input<'d> {
+
+        // Double check that the pin interrupt has been disabled by IRQ handler
+        if self.pin.gpio().icr(self.pin.pin()).read().bits() & (1 << self.pin.pin()) == 0 {
+            if INTERRUPT_DETECTED[self.pin.port()].swap(false, Ordering::Relaxed) {
```
Isn't there a bug here? What if two pins on the same port trigger an interrupt? You could produce `Poll::Ready(())` for the wrong pin. How about changing the `AtomicBool` to an `AtomicU32` where each bit corresponds to the matching pin number?

Then you would check if a given bit would be set here.

```rust
let interrupt = INTERRUPT_DETECTED[self.pin.port()].load(Ordering::Relaxed);
if interrupt & (1 << self.pin.pin()) != 0 {
    interrupt &= !(1 << self.pin.pin());
    INTERRUPT_DETECTED[self.pin.port()].store(interrupt, Ordering::Relaxed);
    // ...
```

MathisDeroo replied:
Indeed good point !! I'll modify to replace the AtomicBool to AtomicU32 that will match the pin number within the PORT

MathisDeroo committed:
Modify INTERRUPT_DETECTED (AtomicBool to AtomicU32) to work with pin number and not only port number interrupt

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo committed:
add embedded_hal_async::digital::* traits

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

felipebalbi commented on src/gpio.rs:
```diff
@@ -3,13 +3,113 @@
+        }
+
+        INTERRUPT_DETECTED[port_index].fetch_or(isfr, Ordering::Relaxed);
+    }
```
because of the `BitIter` usage, we can modify this slightly:

```suggestion
    for pin in BitIter(isfr) {
        // Clear all pending interrupts
        gpio.isfr0().write(|w| unsafe { w.bits(1 << pin) });
        gpio.icr(pin).modify(|_, w| w.irqc().irqc0()); // Disable interrupt

        INTERRUPT_DETECTED[port_index].fetch_or(isfr, Ordering::Relaxed);

        // Wake the corresponding port waker
        if port_index < WAKERS.len() {
            WAKERS[port_index].wake();
        }
    }
```

MathisDeroo replied:
Done, added commit. Thanks for the suggestion

MathisDeroo committed:
Update irq_handler with BitIter loop

Co-authored-by: Felipe Balbi <felipe@balbi.sh>

felipebalbi reviewed (approved):
I think this is good to go in. Anything else that might be left can be improved along the way.

felipebalbi committed:
Merge branch 'main' into gpio_interrupt_support

MathisDeroo commented:
> I think this is good to go in. Anything else that might be left can be improved along the way.

Thanks ! @jamesmunns do you have any feedback or can you approve the PR as well ?

jamesmunns commented on src/gpio.rs:
```diff
@@ -3,13 +3,109 @@
+        INTERRUPT_DETECTED[port_index].fetch_or(isfr, Ordering::Relaxed);
+
+        // Wake the corresponding port waker
+        if port_index < WAKERS.len() {
```
Could do:

```rust
if let Some(w) = WAKERS.get(port_index) {
    w.wake();
}
```

jamesmunns commented on src/gpio.rs:
```diff
@@ -3,13 +3,109 @@
+
+const PORT_COUNT: usize = 5;
+
+static WAKERS: [AtomicWaker; PORT_COUNT] = [const { AtomicWaker::new() }; PORT_COUNT];
```
Just noting that this only gives us one waker per port, that means if someone waits for two pins on the same port, they will be subject to "waker churn": the two futures will constantly wake each other and waste CPU time.

The alternatives to this are:

* Live with it
* Make WAKERS be `[[AtomicWaker; 32]; PORT_COUNT]`: uses a fair bit more static ram space
* Use something like `maitake-sync::WaitQueue` or `maitake-sync::WaitMap` to use intrusive queues that can hold more than one waker

We can punt on this decision for now, just wanted to make sure everyone was aware of it. CC @felipebalbi

MathisDeroo replied:
Agree

felipebalbi replied:
if we go with per pin wakers, then we allocate them during `Instance` impl.

jamesmunns replied:
I'll break this off into a follow-up issue.

jamesmunns commented on src/gpio.rs:
```diff
@@ -691,10 +817,92 @@ impl<'d> Input<'d> {
+
+        // Double check that the pin interrupt has been disabled by IRQ handler
+        if self.pin.gpio().icr(self.pin.pin()).read().bits() & (1 << self.pin.pin()) == 0 {
+            if (INTERRUPT_DETECTED[self.pin.port()].fetch_and(!(1u32 << self.pin.pin()), Ordering::Relaxed) & (1u32 << self.pin.pin())) != 0 {
```
Maybe pull out the mask logic so it isn't repeated?

Also we don't always need to suffix with integer types in Rust like in C, there's no implicit casting to worry about :)

```suggestion
            let mask = 1 << self.pin.pin();
            if (INTERRUPT_DETECTED[self.pin.port()].fetch_and(!mask, Ordering::Relaxed) & mask) != 0 {
```

MathisDeroo committed:
Add suggested changes

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo commented:
@felipebalbi when you merged the main branch into this one, in the interrupt.rs, LPUART2 is listed twice in embassy_hal_internal::interrupt_mod!. Refer to failing checks below

felipebalbi committed:
Merge branch 'main' into gpio_interrupt_support

felipebalbi commented:
> @felipebalbi when you merged the main branch into this one, in the interrupt.rs, LPUART2 is listed twice in embassy_hal_internal::interrupt_mod!. Refer to failing checks below

fixed

Commit:
cargo fmt

Signed-off-by: Felipe Balbi <febalbi@microsoft.com>

jamesmunns commented:
Just noting that @MathisDeroo is working on https://github.com/OpenDevicePartnership/embassy-mcxa/issues/49 and plans to include it in this PR before we merge.

MathisDeroo committed:
WIP: Modify Wakers from AtomicWaker to WaitMap, with pin number (per PORT) as key

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo commented:
> Just noting that @MathisDeroo is working on #49 and plans to include it in this PR before we merge.

WIP commit pushed just above. Please have a look and provide your feedback 😄

jamesmunns committed:
Tweak maitake-sync usage

jamesmunns committed:
Improve docs

jamesmunns commented:
Made suggestions in https://github.com/MathisDeroo/embassy-mcxa/pull/1 (still wip)

jamesmunns committed:
refactor a bit

jamesmunns committed:
Move all of the async+interrupt stuff into a module

MathisDeroo committed:
Merge pull request #1 from jamesmunns/gpio_interrupt_support

Tweak maitake-sync usage

MathisDeroo committed:
Remove defmt debug traces

Signed-off-by: Mathis Deroo <mathis.deroo@nxp.com>

MathisDeroo commented:
Added @jamesmunns suggestion for the interrupt handling using WaitMap. We both tested on HW and it works well. Thanks again @jamesmunns, please approve the PR if needed

Commit:
cargo vet

jamesmunns committed:
Move e-hal-async impls into the gated block

jamesmunns committed:
"rt", begone!

jamesmunns commented:
@MathisDeroo I spoke with @felipebalbi, and we realized that the "rt" feature gating was a little silly, as it's fairly rare in practice to ever actually *want* that.

in https://github.com/MathisDeroo/embassy-mcxa/pull/2 I ended up removing the feature entirely, but wanted to keep you in the loop!
